File: src/espn_boxscore.ts
// src/espn_boxscore.ts
// ESPN undocumented API: scoreboard by date, summary (box score) by gameId. No key required.

import fetch from "node-fetch";

const ESPN_SCOREBOARD =
  "https://site.web.api.espn.com/apis/site/v2/sports/basketball/nba/scoreboard";
const ESPN_SUMMARY =
  "https://site.web.api.espn.com/apis/site/v2/sports/basketball/nba/summary";

const RATE_DELAY_MS = 1000;

function delay(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

export interface PlayerGameStats {
  points: number;
  rebounds: number;
  assists: number;
  threePointFieldGoalsMade: number;
  [key: string]: number;
}

/** Fetch scoreboard for date (YYYY-MM-DD or YYYYMMDD). Returns game/event IDs. */
export async function getScoreboardGameIds(date: string): Promise<string[]> {
  const datesParam = date.replace(/-/g, "");
  const url = `${ESPN_SCOREBOARD}?dates=${datesParam}`;
  const res = await fetch(url, {
    headers: { Accept: "application/json", "User-Agent": "NBA-Props-Optimizer/1.0" },
  });
  if (!res.ok) return [];
  const data = (await res.json()) as { events?: { id: string }[] };
  const events = data.events ?? [];
  return events.map((e) => e.id).filter(Boolean);
}

function onePlayerStats(raw: Record<string, unknown>): PlayerGameStats {
  const stats: PlayerGameStats = {
    points: 0,
    rebounds: 0,
    assists: 0,
    threePointFieldGoalsMade: 0,
    steals: 0,
    blocks: 0,
    turnovers: 0,
  };

  // ESPN summary returns stats in two formats:
  //   1. boxscore.teams[].statistics[].athletes[].stats[] (position-row CSV values)
  //   2. boxscore.players[].statistics.categories[].athletes[].stats[] (same CSV)
  // The CSV row order follows the column "labels" array at the category level.
  // We also handle the named {name, value} format used by some ESPN endpoints.

  const statList = raw.statistics ?? raw.stats;
  if (Array.isArray(statList)) {
    for (const s of statList) {
      const rec = s as Record<string, unknown>;
      const name = ((rec.name ?? rec.label ?? rec.abbreviation ?? "") as string).toLowerCase();
      const val = parseFloat(String(rec.displayValue ?? rec.value ?? 0)) || 0;

      if ((name.includes("point") || name === "pts") && !name.includes("three") && !name.includes("3")) stats.points = val;
      else if (name.includes("rebound") || name === "reb") stats.rebounds = val;
      else if (name.includes("assist") || name === "ast") stats.assists = val;
      else if (name.includes("three") || name === "3pt" || name === "3pm" || name === "3ptm") stats.threePointFieldGoalsMade = val;
      else if (name.includes("steal") || name === "stl") stats.steals = val;
      else if (name.includes("block") || name === "blk") stats.blocks = val;
      else if (name.includes("turnover") || name === "to" || name === "tov") stats.turnovers = val;
    }
  }

  return stats;
}

/** ESPN header label → PlayerGameStats field key. */
const LABEL_TO_FIELD: Record<string, keyof PlayerGameStats> = {
  min: "minutes" as any,
  pts: "points",
  reb: "rebounds",
  ast: "assists",
  stl: "steals",
  blk: "blocks",
  to: "turnovers",
  "3pm": "threePointFieldGoalsMade",
};

/**
 * Parse ESPN summary boxscore into player -> stats.
 *
 * ESPN v2 summary format: boxscore.players[].statistics[].{labels, athletes[]}
 * where each athlete has a `stats` array of CSV values matching `labels` order.
 * Fallback: boxscore.teams[].athletes[] with named statistics.
 */
function parseBoxScorePlayers(summary: unknown): Map<string, PlayerGameStats> {
  const out = new Map<string, PlayerGameStats>();
  const box = (summary as { boxscore?: unknown }).boxscore;
  if (!box || typeof box !== "object") return out;

  // Strategy 1: boxscore.players[].statistics[].{labels, athletes[].stats[]}
  const playersArr = (box as any).players;
  if (Array.isArray(playersArr)) {
    for (const teamGroup of playersArr) {
      const statsCategories = teamGroup.statistics;
      if (!Array.isArray(statsCategories)) continue;
      for (const cat of statsCategories) {
        const labels: string[] = (cat.labels ?? cat.keys ?? []).map((l: string) => l.toLowerCase());
        const athletes = cat.athletes;
        if (!Array.isArray(athletes) || labels.length === 0) continue;
        for (const ath of athletes) {
          const athRec = ath as Record<string, unknown>;
          const athleteObj = athRec.athlete as Record<string, unknown> | undefined;
          const displayName = (
            (athleteObj?.displayName ?? athleteObj?.name ?? athleteObj?.shortName) ??
            (athRec.displayName ?? athRec.name ?? athRec.shortName) ?? ""
          ) as string;
          if (!displayName) continue;

          const statsArr = athRec.stats;
          if (!Array.isArray(statsArr)) continue;

          const playerKey = normalizePlayerName(displayName);
          const existing = out.get(playerKey) ?? {
            points: 0, rebounds: 0, assists: 0,
            threePointFieldGoalsMade: 0, steals: 0, blocks: 0, turnovers: 0,
          };

          for (let i = 0; i < labels.length && i < statsArr.length; i++) {
            const label = labels[i];
            const field = LABEL_TO_FIELD[label];
            if (!field) continue;
            // Stats can be "5" or "1-3" (FG format) — take raw numeric value
            const raw = String(statsArr[i]);
            const val = parseFloat(raw) || 0;
            (existing as any)[field] = val;
          }
          out.set(playerKey, existing);
        }
      }
    }
  }
  if (out.size > 0) return out;

  // Strategy 2: boxscore.teams[].athletes[] (older format)
  const teams = (box as { teams?: { athletes?: unknown[] }[] }).teams;
  if (Array.isArray(teams)) {
    for (const team of teams) {
      const athletes = team.athletes;
      if (!Array.isArray(athletes)) continue;
      for (const a of athletes) {
        const raw = a as Record<string, unknown>;
        const displayName = (raw.displayName ?? raw.name ?? raw.shortName ?? "") as string;
        if (!displayName) continue;
        out.set(normalizePlayerName(displayName), onePlayerStats(raw));
      }
    }
  }
  if (out.size > 0) return out;

  // Strategy 3: recursive walk fallback
  walkForAthletes(box as Record<string, unknown>, out);
  return out;
}

function walkForAthletes(obj: Record<string, unknown>, out: Map<string, PlayerGameStats>): void {
  const displayName = (obj.displayName ?? obj.name ?? obj.shortName ?? "") as string;
  const hasStats = Array.isArray(obj.statistics ?? obj.stats);
  if (displayName && hasStats) {
    out.set(normalizePlayerName(displayName), onePlayerStats(obj));
    return;
  }
  for (const v of Object.values(obj)) {
    if (v && typeof v === "object" && !Array.isArray(v)) walkForAthletes(v as Record<string, unknown>, out);
    else if (Array.isArray(v)) for (const item of v) if (item && typeof item === "object") walkForAthletes(item as Record<string, unknown>, out);
  }
}

/** Normalize for matching: lowercase, collapse spaces, remove Jr./III/etc. */
function normalizePlayerName(name: string): string {
  return name
    .replace(/\s*(Jr\.?|III?|IV|II)\s*$/i, "")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

/** Fetch summary for one game; return player -> stats. */
export async function getBoxScoreForGame(gameId: string): Promise<Map<string, PlayerGameStats>> {
  const url = `${ESPN_SUMMARY}?event=${gameId}`;
  const res = await fetch(url, {
    headers: { Accept: "application/json", "User-Agent": "NBA-Props-Optimizer/1.0" },
  });
  if (!res.ok) return new Map();
  const summary = (await res.json()) as unknown;
  return parseBoxScorePlayers(summary);
}

/** Get all player stats for a date (scoreboard + one summary per game with 1s delay). One player appears in at most one game. */
export async function fetchAllPlayerStatsForDate(
  date: string
): Promise<Map<string, PlayerGameStats>> {
  const gameIds = await getScoreboardGameIds(date);
  const allStats = new Map<string, PlayerGameStats>();
  for (const gameId of gameIds) {
    await delay(RATE_DELAY_MS);
    const gameStats = await getBoxScoreForGame(gameId);
    for (const [name, stats] of gameStats) allStats.set(name, stats);
  }
  return allStats;
}

/** Map our stat key to ESPN box value. Supports all NBA categories + combos. */
export function getStatValueFromBox(stats: PlayerGameStats, stat: string): number {
  const s = stat.toLowerCase().replace(/[\s_-]+/g, "");

  // Single stats
  if ((s.includes("point") || s === "pts") && !s.includes("3") && !s.includes("three") && !s.includes("rebound") && !s.includes("assist"))
    return stats.points ?? 0;
  if (s.includes("rebound") || s === "reb") return stats.rebounds ?? 0;
  if (s.includes("assist") || s === "ast") return stats.assists ?? 0;
  if (s.includes("three") || s === "3pm" || s === "3pt" || s === "threes" || s === "threesmade" || s === "threepointersmade")
    return stats.threePointFieldGoalsMade ?? 0;
  if (s.includes("steal") || s === "stl") return stats.steals ?? 0;
  if (s.includes("block") || s === "blk") return stats.blocks ?? 0;
  if (s.includes("turnover") || s === "to" || s === "tov") return stats.turnovers ?? 0;

  // Combo stats
  if (s === "pra" || s === "pointsreboundsassists" || s === "pts+reb+ast")
    return (stats.points ?? 0) + (stats.rebounds ?? 0) + (stats.assists ?? 0);
  if (s === "pr" || s === "pointsrebounds" || s === "pts+reb")
    return (stats.points ?? 0) + (stats.rebounds ?? 0);
  if (s === "pa" || s === "pointsassists" || s === "pts+ast")
    return (stats.points ?? 0) + (stats.assists ?? 0);
  if (s === "ra" || s === "reboundsassists" || s === "reb+ast")
    return (stats.rebounds ?? 0) + (stats.assists ?? 0);
  if (s === "stocks" || s === "stealsblocks" || s === "stl+blk")
    return (stats.steals ?? 0) + (stats.blocks ?? 0);

  return 0;
}

File: scripts/espn_boxscore.py
#!/usr/bin/env python3
"""
Fetch NBA box scores from ESPN (no API key). Used by settle_results.py to get
actual stats per player/game for settling leg outcomes.

Usage:
    from scripts.espn_boxscore import fetch_all_player_stats_for_date
    stats = fetch_all_player_stats_for_date("2026-03-06")  # player_key -> {points, rebounds, ...}
    value = get_stat_value_from_box(stats.get("jalen brunson"), "points")
"""

import re
import time
from typing import Any

import requests

ESPN_SCOREBOARD = "https://site.web.api.espn.com/apis/site/v2/sports/basketball/nba/scoreboard"
ESPN_SUMMARY = "https://site.web.api.espn.com/apis/site/v2/sports/basketball/nba/summary"
RATE_DELAY_SEC = 1.0
HEADERS = {"Accept": "application/json", "User-Agent": "NBA-Props-Optimizer/1.0"}


def _normalize_name(name: str) -> str:
    if not name:
        return ""
    s = re.sub(r"\s*(Jr\.?|III?|IV|II)\s*$", "", name, flags=re.IGNORECASE)
    s = " ".join(s.split()).strip().lower()
    return s


def get_scoreboard_game_ids(date: str) -> list[str]:
    """Date as YYYY-MM-DD or YYYYMMDD. Returns list of ESPN event IDs."""
    dates_param = date.replace("-", "")[:8]
    url = f"{ESPN_SCOREBOARD}?dates={dates_param}"
    try:
        r = requests.get(url, headers=HEADERS, timeout=15)
        r.raise_for_status()
        data = r.json()
    except Exception:
        return []
    events = data.get("events") or []
    return [e["id"] for e in events if e.get("id")]


def _one_player_stats(raw: dict[str, Any]) -> dict[str, float]:
    out = {
        "points": 0.0,
        "rebounds": 0.0,
        "assists": 0.0,
        "threePointFieldGoalsMade": 0.0,
        "steals": 0.0,
        "blocks": 0.0,
        "turnovers": 0.0,
    }
    stat_list = raw.get("statistics") or raw.get("stats")
    if not isinstance(stat_list, list):
        return out
    for s in stat_list:
        if not isinstance(s, dict):
            continue
        name = ((s.get("name") or s.get("label") or s.get("abbreviation")) or "").lower()
        try:
            val = float(s.get("displayValue") or s.get("value") or 0)
        except (TypeError, ValueError):
            val = 0
        if "point" in name and "three" not in name and "3" not in name:
            out["points"] = val
        elif "rebound" in name or name == "reb":
            out["rebounds"] = val
        elif "assist" in name or name == "ast":
            out["assists"] = val
        elif "three" in name or name in ("3pt", "3pm", "3ptm"):
            out["threePointFieldGoalsMade"] = val
        elif "steal" in name or name == "stl":
            out["steals"] = val
        elif "block" in name or name == "blk":
            out["blocks"] = val
        elif "turnover" in name or name in ("to", "tov"):
            out["turnovers"] = val
    return out


LABEL_TO_FIELD = {
    "min": "minutes",
    "pts": "points",
    "reb": "rebounds",
    "ast": "assists",
    "stl": "steals",
    "blk": "blocks",
    "to": "turnovers",
    "3pm": "threePointFieldGoalsMade",
}


def _parse_boxscore_players(summary: Any) -> dict[str, dict[str, float]]:
    out: dict[str, dict[str, float]] = {}
    box = (summary or {}).get("boxscore")
    if not box or not isinstance(box, dict):
        return out

    players_arr = box.get("players")
    if isinstance(players_arr, list):
        for team_group in players_arr:
            for cat in team_group.get("statistics") or []:
                labels = [str(x).lower() for x in (cat.get("labels") or cat.get("keys") or [])]
                athletes = cat.get("athletes") or []
                if not labels or not athletes:
                    continue
                for ath in athletes:
                    if not isinstance(ath, dict):
                        continue
                    athlete_obj = ath.get("athlete") or {}
                    display_name = (
                        athlete_obj.get("displayName")
                        or athlete_obj.get("name")
                        or athlete_obj.get("shortName")
                        or ath.get("displayName")
                        or ath.get("name")
                        or ath.get("shortName")
                        or ""
                    )
                    if not display_name:
                        continue
                    key = _normalize_name(display_name)
                    existing = out.setdefault(
                        key,
                        {
                            "points": 0,
                            "rebounds": 0,
                            "assists": 0,
                            "threePointFieldGoalsMade": 0,
                            "steals": 0,
                            "blocks": 0,
                            "turnovers": 0,
                        },
                    )
                    stats_arr = ath.get("stats")
                    if isinstance(stats_arr, list):
                        for i, label in enumerate(labels):
                            if i >= len(stats_arr):
                                break
                            field = LABEL_TO_FIELD.get(label)
                            if not field:
                                continue
                            try:
                                existing[field] = float(stats_arr[i])
                            except (TypeError, ValueError):
                                pass
                    out[key] = existing

    if out:
        return out

    for team in box.get("teams") or []:
        for a in team.get("athletes") or []:
            if not isinstance(a, dict):
                continue
            display_name = a.get("displayName") or a.get("name") or a.get("shortName") or ""
            if display_name:
                out[_normalize_name(display_name)] = _one_player_stats(a)
    return out


def get_box_score_for_game(game_id: str) -> dict[str, dict[str, float]]:
    """Fetch one game summary; return player_key -> stats dict."""
    url = f"{ESPN_SUMMARY}?event={game_id}"
    try:
        r = requests.get(url, headers=HEADERS, timeout=15)
        r.raise_for_status()
        return _parse_boxscore_players(r.json())
    except Exception:
        return {}


def fetch_all_player_stats_for_date(date: str) -> dict[str, dict[str, float]]:
    """
    Fetch all player stats for an NBA date. Returns dict: normalized_player_name -> stats.
    Stats include points, rebounds, assists, threePointFieldGoalsMade, steals, blocks, turnovers.
    """
    game_ids = get_scoreboard_game_ids(date)
    all_stats: dict[str, dict[str, float]] = {}
    for gid in game_ids:
        time.sleep(RATE_DELAY_SEC)
        game_stats = get_box_score_for_game(gid)
        for k, v in game_stats.items():
            all_stats[k] = v
    return all_stats


def get_stat_value_from_box(stats: dict[str, float] | None, stat_type: str) -> float:
    """
    Map our stat_type (points, rebounds, assists, threes, etc.) to the value from box stats.
    Combo stats (PRA, PR, etc.) are summed. Returns 0 if unknown or missing.
    """
    if not stats:
        return 0.0
    s = (stat_type or "").lower().replace(" ", "").replace("_", "").replace("-", "")
    pts = stats.get("points", 0) or 0
    reb = stats.get("rebounds", 0) or 0
    ast = stats.get("assists", 0) or 0
    threes = stats.get("threePointFieldGoalsMade", 0) or 0
    stl = stats.get("steals", 0) or 0
    blk = stats.get("blocks", 0) or 0
    tov = stats.get("turnovers", 0) or 0

    if ("point" in s or s == "pts") and "3" not in s and "three" not in s and "reb" not in s and "ast" not in s:
        return float(pts)
    if "rebound" in s or s == "reb":
        return float(reb)
    if "assist" in s or s == "ast":
        return float(ast)
    if "three" in s or s in ("3pm", "3pt", "threes", "threesmade", "threepointersmade"):
        return float(threes)
    if "steal" in s or s == "stl":
        return float(stl)
    if "block" in s or s == "blk":
        return float(blk)
    if "turnover" in s or s in ("to", "tov"):
        return float(tov)
    if s in ("pra", "pointsreboundsassists", "pts+reb+ast"):
        return float(pts + reb + ast)
    if s in ("pr", "pointsrebounds", "pts+reb"):
        return float(pts + reb)
    if s in ("pa", "pointsassists", "pts+ast"):
        return float(pts + ast)
    if s in ("ra", "reboundsassists", "reb+ast"):
        return float(reb + ast)
    if s in ("stocks", "stealsblocks", "stl+blk"):
        return float(stl + blk)
    return 0.0


def find_player_stats(date_stats: dict[str, dict[str, float]], player_name: str) -> dict[str, float] | None:
    """Match leg player name to box score key (normalized; fallback last name)."""
    norm = _normalize_name(player_name or "")
    if not norm:
        return None
    if norm in date_stats:
        return date_stats[norm]
    last = norm.split()[-1] if norm.split() else ""
    for key, st in date_stats.items():
        if key.endswith(" " + last) or key == last:
            return st
    return None
